router/usecase/optimized_routes_test.go

```diff
@@ -674,6 +674,118 @@ func (s *RouterTestSuite) TestGetCustomQuote_GetCustomDirectQuote_Mainnet_UOSMOU
 	s.validateExpectedPoolIDOneRouteOneHopQuote(quote, expectedPoolID)
 }
 
+// TODO
+func (s *RouterTestSuite) TestGetCustomQuote_GetCustomDirectQuotes_Mainnet_UOSMOUSDC() {
+	config := routertesting.DefaultRouterConfig
+	config.MaxPoolsPerRoute = 5
+	config.MaxRoutes = 10
+
+	var (
+		amountIn = osmomath.NewInt(5000000)
+	)
+
+	mainnetState := s.SetupMainnetState()
+
+	// Setup router repository mock
+	routerRepositoryMock := routerrepo.New()
+	routerRepositoryMock.SetTakerFees(mainnetState.TakerFeeMap)
+
+	// Setup pools usecase mock.
+	poolsUsecase := poolsusecase.NewPoolsUsecase(&domain.PoolsConfig{}, "node-uri-placeholder", routerRepositoryMock, domain.UnsetScalingFactorGetterCb)
+	poolsUsecase.StorePools(mainnetState.Pools)
+
+	routerUsecase := routerusecase.NewRouterUsecase(routerRepositoryMock, poolsUsecase, config, emptyCosmWasmPoolsRouterConfig, &log.NoOpLogger{}, cache.New(), cache.New())
+
+	// Test cases
+	testCases := []struct {
+		// test name
+		name string
+
+		// token being swapped
+		tokenIn sdk.Coin
+
+		// token to be received
+		tokenOutDenom string
+
+		// pools route path for swap
+		poolID []uint64
+
+		// usually it's the number of pools given,
+		// unless any of those pools does not have given asset pair.
+		expectedNumOfRoutes int
+
+		// for single-hop it matches poolID slice
+		expectedPoolID []uint64
+
+		err error
+	}{
+		{
+			name:          "Single pool: OSMO-ATOM - happy case",
+			tokenIn:       sdk.NewCoin(UOSMO, amountIn),
+			tokenOutDenom: ATOM,
+			poolID: []uint64{
+				1, // OSMO - ATOM
+			},
+			expectedNumOfRoutes: 1,
+			expectedPoolID:      []uint64{1},
+		},
+		{
+			name:          "Single pool: OSMO-ATOM - fail case: out denom not found",
+			tokenIn:       sdk.NewCoin(UOSMO, amountIn),
+			tokenOutDenom: ATOM,
+			poolID: []uint64{
+				1093, // OSMO - AKT
+			},
+			err:                 routerusecase.ErrTokenOutDenomPoolNotFound,
+		},
+		{
+			name:          "Single pool: ATOM-OSMO - fail case: in denom not found",
+			tokenIn:       sdk.NewCoin(ATOM, amountIn),
+			tokenOutDenom: UOSMO,
+			poolID: []uint64{
+				1480, // AKT - USDC
+			},
+			err:                 routerusecase.ErrTokenInDenomPoolNotFound,
+		},
+		{
+			name:          "Multi pool: OSMO-USDC - happy case",
+			tokenIn:       sdk.NewCoin(UOSMO, amountIn),
+			tokenOutDenom: USDC,
+			poolID: []uint64{
+				1093, // OSMO - AKT
+				1480, // AKT - USDC
+			},
+			expectedNumOfRoutes: 2,
+			expectedPoolID:      []uint64{1093, 1480},
+		},
```

Comment:
Let me construct a made up example that will currently break

```go
tokenIn:       sdk.NewCoin(UOSMO, amountIn),
tokenOutDenom: USDC,
poolID: []uint64{
   5, // OSMO, AKT, ATOM
   24 // AKT - USDC
}
```

In this example, when we complete the iteration over the first pool, how do we know which tokenOutDenom becomes the tokenIn denom for the next iteration?

We need to add another query param `tokenOutDenoms` which is a slice of token out denoms corresponding to the pool ID stemming from `poolIDs` query param in the same position.

Just for context (we should not pursue this) - alternatively, we could try using `GetCandidateRoutes` route that would construct the slice of pool IDs and respective tokenOutDenoms on the path from the user-provided token in denom to the user-provided token out denom.

However, the way this algorithm works is that it finds 20 routes between two tokens and does not search further. In some cases, we might want to investigate routes that this routine cannot find. As a result, we should push the responsibility of selecting the pool IDs and relevant token out denoms to the user.

What I'm thinking about is something like this:
```go
func GetCustomDirectQuoteMultiPool(ctx context.Context, tokenIn sdk.Coin, poolID []uint64, tokenOutDenoms []string) {
   for i, poolID := range poolIDs {
      curTokenOutDenom :=  tokenOutDenoms[i]
      
      // do stuff
      curAmtOut, ... := GetCustomDirectQuote(..., tokenIn, curTokenOutDenom)
     
      tokenIn = sdk.NewCoin(curTokenOutDenom, curAmtOut)
   }
}
```

Please let me know if that makes sense ⏫

Reply:
That totally makes sense and explains everything, thanks for taking your time! I was asking the very same question myself while working on the task since your last [comment](https://github.com/osmosis-labs/sqs/pull/284#discussion_r1628725802), but somehow your suggestion there slipped through my eyes which led previous version of the solution. Marking this thread as resolved for now.